spec/sidekiq_flow/adapters/set_storage_spec.rb

```diff
@@ -0,0 +1,497 @@
+      let(:task) { workflow.tasks.first }
+
+      before do
+        described_class.start_workflow(workflow)
+      end
+
+      it 'should store task info in redis' do
+        expect {
+          task.fail!
+          subject
+        }.to change {
+          SidekiqFlow::Client.find_task(workflow.id, task.class.name).status
+        }.from('enqueued').to('failed')
+      end
+
+      context 'NOT yet succeeded workflow' do
+        it 'should move workflow key to finished set' do
+          subject
+          expect( redis.sismember(IN_PROGRESS, SidekiqFlow::Client.find_workflow_key(workflow.id)) ).to eq(true)
+        end
+      end
+    end
+  end
+
+  describe '.find_workflow' do
+    subject { described_class.find_workflow(workflow.id) }
+
+    context 'success' do
+      before do
+        described_class.start_workflow(workflow)
+      end
+
+      it 'should return a Workflow instance' do
+        result = subject
+        expect(result).to be_a_kind_of(TestWorkflow)
+      end
+
+      it 'should return a the correct Workflow instance' do
+        result = subject
+        expect(result.id).to eq(workflow.id)
+      end
+    end
+
+    context 'exception' do
+      it 'should raise WorkflowNotFound' do
+        expect {
+          subject
+        }.to raise_exception(SidekiqFlow::WorkflowNotFound)
+      end
+    end
+  end
+
+  describe '.destroy_workflow' do
+    subject { described_class.destroy_workflow(workflow.id) }
+
+    context 'success' do
+      before do
+        described_class.start_workflow(workflow)
+      end
+
+      it 'should delete the workflow' do
+        expect {
+          subject
+        }.to change { redis.keys.count }.from(2).to(1)
+
+        expect( redis.sismember(FINISHED, SidekiqFlow::Client.find_workflow_key(workflow.id)) ).to eq(false)
+      end
+    end
+  end
+
+  describe '.destroy_succeeded_workflows' do
+    let(:workflow2) { TestWorkflow.new(id: 345) }
+
+    subject { described_class.destroy_succeeded_workflows }
+
+    context 'success' do
+      before do
+        described_class.start_workflow(workflow)
+        SidekiqFlow::Worker.drain
+        described_class.start_workflow(workflow2)
+      end
+
+      it 'should ONLY delete succeeded workflows' do
+        expect( redis.sismember(IN_PROGRESS, SidekiqFlow::Client.find_workflow_key(workflow2.id)) ).to eq(true)
+        expect( redis.sismember(FINISHED, SidekiqFlow::Client.find_workflow_key(workflow.id)) ).to eq(true)
+
+        subject
+
+        expect( redis.sismember(IN_PROGRESS, SidekiqFlow::Client.find_workflow_key(workflow2.id)) ).to eq(true)
+        expect( redis.sismember(FINISHED, SidekiqFlow::Client.find_workflow_key(workflow.id)) ).to eq(false)
+      end
+
+      it 'should change redis key count' do
+        expect {
+          subject
+        }.to change {
+          redis.keys.count
+        }.from(4).to(2)
+      end
+
+      it 'should NOT find deleted workflows' do
```

Comment:
break into 2:

`it should raise error when looking for deleted workflows`
`it should NOT raise an error when looking for in-progress workflows`